Proj 6 coordinate operation handling -- report errors and warnings to users

nyalldawson opened the pull request:
This PR extends the Proj 6 support available in master with methods to allow warnings and errors from proj to be pushed back and displayed to users. This is critical functionality -- by hiding these warnings and errors we allow users to accidentally create nonsensical or invalid coordinate transformations silently. Better to push the warnings into their faces so that they can make informed decisions based on this.

In particular,  warnings are shown when:

- A coordinate transform cannot be created, e.g. due to not making sense (transforming earth coords to mars coords)
- A coordinate transform cannot be created due to a missing grid file (along with links to download the grid if available)
- A coordinate transform was created, BUT a better one is possible, it just is missing some prerequisite on the user's system (again, with links to download if available)

Sponsored by ICSM

nyalldawson committed:
Remove unused setFinder method

nyalldawson committed:
Refactor to extract useful method

nyalldawson committed:
Add mechanisms to set handlers for coordinate transform creation errors and warnings

nyalldawson committed:
Always attempt to create the preferred Proj coordinate operation
for transforms, even if grids are not available.

And then use custom handlers or message log to warn users that
better operations are possible, just not available on their system.

Also use similar approach to raise errors when either no operation
is possible between two coordinate systems, or only ONE is yet
it is not available on the current system (e.g. due to missing
grid file)

Sponsored by ICSM

nyalldawson committed:
Don't show generic proj error message to users

nyalldawson committed:
Show nice friendly warnings to users when

- A coordinate transform cannot be created, e.g. due to not making
sense (transforming earth coords to mars coords)
- A coordinate transform cannot be created due to a missing grid
file (along with links to download the grid if available)
- A coordinate transform was created, BUT a better one is possible,
it just is missing some prerequisite on the user's system (again,
with links to download if available)

Sponsored by ICSM

nyalldawson committed:
Allow clickable links in QgsMessageViewer

nyalldawson commented:
Screencast:

![Peek 2019-05-31 13-55](https://user-images.githubusercontent.com/1829991/58680614-d6543580-83ab-11e9-92c9-47e8fac159e7.gif)

nirvn commented:
@nyalldawson , nice. Out of curiosity from looking at the GIF, what does "an accuracy of -1 [i.e. negative one] meter" means?

nyalldawson committed:
Update datum transform widgets and dialog for proj 6

nyalldawson commented:
@nirvn fixed -- it means "unknown"

nyalldawson committed:
Fix accuracy message when operation accuracy is unknown

nyalldawson committed:
Don't select not-available operations by default

nyalldawson committed:
Fix caching of coordinate transforms

nyalldawson committed:
Fix doxygen

wonder-sk commented:
wow - this is really great

nyalldawson committed:
Fix application of default datum transform on proj 6 builds

nyalldawson committed:
Update test results for proj 6

nyalldawson committed:
Whenever calculating proj coordinate operation strings, ALWAYS
base them on normalized operations

QGIS doesn't need (or want) to care about strictly following
projection definitions here, we ALWAYS want X/Y to be X/Y

nyalldawson committed:
Don't allow building QGIS on proj 6.0 versions

Require at least 6.1.0. We need proj_normalize_for_visualization,
and the workarounds for building without it are too extensive,
too fragile, and result in too much spaghetti code.

Gotta make the hard call here ;)

nyalldawson committed:
Update test results

nyalldawson committed:
Add method to retrieve grid details from a proj string

nyalldawson committed:
Add friendly, descriptive errors when opening a project which has
preset non-default coordinate operations set between a crs pair,
which use grids that are not available on the local system

nyalldawson commented:
IMO this completes the epic QGIS -> proj 6 upgrade. May I well and truly be retired before the next one ;)

nirvn commented:
:tada:

nirvn commented on src/core/qgscoordinatetransform.cpp:
```diff
@@ -652,7 +652,18 @@ void QgsCoordinateTransform::transformCoords( int numPoints, double *x, double *
+                      z, sizeof( double ), numPoints,
+                      nullptr, sizeof( double ), 0 );
+  projResult = proj_errno( projData );
+  // ewww - this logic is gross. We should drop support for PROJ 6.0 as quickly as possible and dump this code (in favour of built in methods used for >=6.1 builds)
```
I believe you can get rid of the eww comment here.

nyalldawson replied:
Thanks, fixed

nyalldawson committed:
Remove outdated comment

nyalldawson commented:
Woohoo!

rduivenvoorde commented:
Thanks @nyalldawson  !!

PedroVenancio commented:
Really really great stuff @nyalldawson !!!

Thank you very much to you, ICSM and PROJ/GDAL guys!!!

Can the warn message show where the grid shift files should be manually pasted (proj folder), when they are missing? This can be useful for users.
